orderer/common/filerepo/filerepo.go

```diff
@@ -0,0 +1,169 @@
+		return nil, err
+	}
+
+	fileRepoDir := filepath.Join(repoParentDir, fileSuffix)
+
+	// Make the file repo if it doesn't already exist
+	if _, err := fileutil.CreateDirIfMissing(fileRepoDir); err != nil {
+		return nil, err
+	}
+
+	// Sync parent dir after making file repo dir
+	if err := fileutil.SyncDir(repoParentDir); err != nil {
+		return nil, err
+	}
+
+	files, err := ioutil.ReadDir(fileRepoDir)
+	if err != nil {
+		return nil, err
+	}
+
+	// Remove existing transient files in the repo
+	for _, f := range files {
+		transientFilePattern := "*" + fileSuffix + "~"
+		isTransientFile, err := filepath.Match(transientFilePattern, f.Name())
+		if err != nil {
+			return nil, err
+		}
+		if isTransientFile {
+			os.Remove(filepath.Join(fileRepoDir, f.Name()))
+		}
+	}
+
+	if err := fileutil.SyncDir(fileRepoDir); err != nil {
+		return nil, err
+	}
+
+	return &Repo{
+		transientFileMarker: "~",
+		fileSuffix:          fileSuffix,
+		fileRepoDir:         fileRepoDir,
+	}, nil
+}
+
+// Save atomically persists the content to suffix/baseName+suffix file by first writing it
+// to a tmp file marked by the transientFileMarker and then moves the file to the final
+// destination indicated by the FileSuffix.
+func (r *Repo) Save(baseName string, content []byte) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	fileName := fmt.Sprintf("%s.%s", baseName, r.fileSuffix)
+	dest := filepath.Join(r.fileRepoDir, fileName)
+
+	if _, err := os.Stat(dest); err == nil {
+		return errors.Errorf("file already exists at %s", dest)
+	}
+
+	tmpFileName := fileName + r.transientFileMarker
+	if err := fileutil.CreateAndSyncFileAtomically(r.fileRepoDir, tmpFileName, fileName, content, repoFilePermPrivateRW); err != nil {
+		return err
+	}
+
+	return fileutil.SyncDir(r.fileRepoDir)
+}
+
+// Remove removes the file associated with baseName from the file system.
+func (r *Repo) Remove(baseName string) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	fileName := fmt.Sprintf("%s.%s", baseName, r.fileSuffix)
+	filePath := filepath.Join(r.fileRepoDir, fileName)
+
+	if err := os.RemoveAll(filePath); err != nil {
+		return err
+	}
+
+	return fileutil.SyncDir(r.fileRepoDir)
+}
+
+// Read reads the file in the fileRepo associated with baseName's contents.
+func (r *Repo) Read(baseName string) ([]byte, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	fileName := fmt.Sprintf("%s.%s", baseName, r.fileSuffix)
+	filePath := filepath.Join(r.fileRepoDir, fileName)
+	return ioutil.ReadFile(filePath)
+}
+
+// List parses the directory and produce a list of file names, filtered by suffix.
+func (r *Repo) List() ([]string, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	var repoFiles []string
+
+	files, err := ioutil.ReadDir(r.fileRepoDir)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, f := range files {
+		isFileSuffix, err := filepath.Match("*"+r.fileSuffix, f.Name())
+		if err != nil {
+			return nil, err
+		}
+		if isFileSuffix {
+			repoFiles = append(repoFiles, f.Name())
+		}
```

Comment:
Is this necessary? Given that methods on `Repo` are thread-safe and `New()` removes all transientFiles, do we need to perform the match based on fileSuffix. If it is necessary, please update the test to reflect this. Otherwise, remove this part of the code.

Reply:
It's still useful to ensure other files are filtered out if they somehow end up there. The tests already reflect this by filtering out `testdata/joinblock/mychannel.block` and only returning `testdata/joinblock/mychannel.joinblock`

Reply:
Probably, I missed that. If test is there, then it is fine. 

> if they somehow end up there

I do not understand. Are you referring to some attack scenario?